docs/scos/dev/architecture/architecture-performance.md

```diff
@@ -0,0 +1,211 @@
+
+## General Performance Challenges in Architecture Design
+In this part you will find the most architecture design mistakes and impediments in order to fulfil the business requirements with losing the performance.
+
+### Duplications of Slow Operations
+Sometime it’s mandatory to have a slow operation during one transaction because of business requirements. This slow part of functionally might be very small and is hidden behind of an API but the usage of this API can go out of control.
+
+Let’s take one example and explain the impact of a bad architecture design for this: 
+
+Imagine you have a method which called caluculateDiscount() wants to generate some discounts for  a cart items, but this method for each time call takes 100ms and this might be a proper response time for an API. now think of another business requirement which needs to calculate discount for 10 separated group of Items in a cart. Therefore we need to call the caluculateDiscount() 10 times which can lead to 1000ms (1 sec) and this is now a performance problem.
+
+### Duplications of Database Queries
+During the project implementation it’s possible that developers execute similar queries which are returning same result or subset of data from it in one transaction. Architects must make sure that the database interactions set to the lowest possible number and this can be achievable by
+
+Merging several queries to only one query with a bigger result (Unfiltered)
+
+ Aggregate the duplicates query to one query and share the result to the stack of the code execution (Memory)
+
+{% info_block warningBox %}
+
+Memory leaks must be carefully checked during query optimizations as the results will be bigger or shared in one transaction.
+
+{% endinfo_block %}
+
+
+### Optimistic vs. Pessimistic Locking
+It’s very possible that developers use explicit locks to prevent the race conditions or other problems, but this can easily impact the performance in high load of traffic. All requests now needs to wait for the lock and this will turn the parallel request processing to Sequentional processing and can increase the response time of all queued requests.
+
+Some of the Pessimistic locking use cases:
+
+- Concurrent Session problems (Redis, File)
+- Generate Unique Numbers (Database)
+
+Architect can recommend to use Optimistics locking with several different implementations according to the problems.
+
+### Synchronous Communications and 3rd Party Calls
+Another mistake for architecture is to rely on the 3rd party response time in order to achieve a promised performances for the application. Having a direct external call to a 4rd party organisation during a transaction can make the performance unpredictable and will impact the user experience behaviour.
+
+Architects needs to fulfil the requirements by providing a different solution like asynchronous communication or pre/after indirect hooks which are not visible for the end users.
```

Comment:
What does *pre/after indirect hooks* mean here? Could you please rephrase?

Reply:
Uff, I still understand it as a "asynchronous communication". If you need to publish this urgently - let's remove the "or" part, but check with Ehsan what he meant once he is back.